### ISSUE TITLE
Some questions for Python implementation

### ISSUE THREAD

**Issue by collinwebdesigns** [deadash/apple_cache#1]
Thanks @deadash for your working implementation of apple-cache-server. I want to implement this in Python using FastAPI.

I have some questions for my understanding and implementation:
- where do I get the `push-token` in `cache.json` from - or is this a static value?
- where do I get the `guid` in `cache.json` from
- what does "Ensure that all five codes are unified." mean, unified for this cache-server-instance or globally unified? Can anybody reuse your values?!
- all files in `blob/*` are created by the cache-server, so they don't have to be commited to GitHub, or am I wrong?

I'll get started until I receive your answers. Thank you already!

**Comment by deadash** [comment]
1. The `push-token` in `cache.json` is related to Apple's push notification service. It isn't used in the cache server's core functionality, so you can set it to any value. If you want to receive push notifications, you could set it to a valid token, but I haven't tested this aspect, and it doesn't affect the registration process whether it's valid or not.

2. The `guid` is also not strictly required to be a specific value; it can be generated randomly. It is used for identification purposes by the download side, likely to distinguish between multiple services under the same public network. If you need a specific value, you can check the logs on a Mac virtual machine to find the `guid` of your Mac computer and use that.

3. The phrase "Ensure that all five codes are unified" refers to the requirement that the five unique identifiers associated with a Mac computer (like `Gq3489ugfi`) need to be distinct from others. These values must be unique to avoid conflicts with other users. If two users use the same identifiers, one may get disconnected. These identifiers can be obtained from a virtual machine.

4. The files in the `blob/*` directory are indeed created by the cache server and should not be committed to GitHub. You should follow the documentation related to Git Large File Storage (Git LFS) or download the source code package to get these files (refer to the corresponding documentation or website).

5. The `cert.cer` file in the `blob/` directory originates from Apple. There's a URL from which this file can be directly obtained. I'll provide the source later. I included it in the repository to avoid the need to download it separately.

**Comment by collinwebdesigns** [comment]
Thanks for your answers!

Then I would like to know what happens [here](https://github.com/deadash/apple_cache/blob/main/api/src/main.rs#L39) ([`patch::call_0(&mut self.uc, &pcert)`](https://github.com/deadash/apple_cache/blob/main/api/src/patch/emulator.rs#L375-L402)) with the certificate. I don't understand that. Can you explain it to me line by line or write a pseudo code?

**Comment by deadash** [comment]
> Thanks for your answers!
> 
> 
> 
> Then I would like to know what happens [here](https://github.com/deadash/apple_cache/blob/main/api/src/main.rs#L39) ([`patch::call_0(&mut self.uc, &pcert)`](https://github.com/deadash/apple_cache/blob/main/api/src/patch/emulator.rs#L375-L402)) with the certificate. I don't understand that. Can you explain it to me line by line or write a pseudo code?
> 
> 
> 
> 
> 
> 
> 
> 

In the code `patch::call_0(&mut self.uc, &pcert)`, if you are not using an emulator, this part of the code would not execute. Essentially, this code operates the same as in the direct execution mode. The emulation mode involves invoking this function by simulating the process, where the certificate is passed in for initialization, and that's all there is to it.

**Comment by collinwebdesigns** [comment]
> if you are not using an emulator

Ah, I think I got it. I don't have to look at all `emulator` files, instead using the other ones in each module?

**Comment by collinwebdesigns** [comment]
I have now got the `/lcdn/session` process running. Now I can't get any further with the `signing` of the `/lcdn/register`.
Currently I load the `cache.json` and add the node `session-token` with the respective value from the session. The whole thing is now available to me as a string. I also have the response from `/lcdn/session` as a string. I think I have all the pieces of the puzzle and just have to put them together. How does the signing process work?

**Comment by deadash** [comment]
I have completed the Python bindings, and you can pull the code to take a look. If you are on Windows, you can download the [pre-compiled](https://github.com/deadash/apple_cache/releases/download/v0.1.0/AppleCacheRelease-0.1.zip) version and run it to try it out.

**Comment by collinwebdesigns** [comment]
> I have completed the Python bindings, and you can pull the code to take a look.

Thank you, but these are only partially helpful, as they indirectly refer to the Rust code again.

I would like to implement the whole thing cleanly in python. I am currently failing at singing. Can you explain to me in more detail how this works? Unfortunately, I don't understand the Rust code. I can't do much with these instructions either (https://github.com/azenla/AppleCache/tree/main/signature). But it doesn't seem to be too complicated, does it?

**Comment by deadash** [comment]
Unfortunately, I did not obtain the actual part of the algorithm. Instead, I used a simulated approach or ran the signature algorithm directly. The algorithm employs high-intensity obfuscation, making it difficult to revert to a plaintext algorithm. I think using pure Python is infeasible for now. It would only be possible if the algorithm was disassembled into a usable form, which is currently beyond my capabilities.

**Comment by collinwebdesigns** [comment]
> I used a simulated approach or ran the signature algorithm directly

What does this mean "directly", are you using some internal macOS calls?

And how does this work on Windows?

**Comment by deadash** [comment]
> > I used a simulated approach or ran the signature algorithm directly
> 
> What does this mean "directly", are you using some internal macOS calls?
> 
> And how does this work on Windows?

Yes, running directly means executing macOS code; I patched the relevant calls. In fact, the 64-bit assembly code is platform-independent and can be run on macOS, Linux, Windows, etc. The only differences are in file structure and calling conventions. If you look at the code, you'll notice that the calling conventions are not Windows conventions. As for macOS-specific functions, I have patched them to return the needed results, which is also why a Mac serial number is required. I directly return them to avoid calling Mac functions to obtain the information.

**Comment by collinwebdesigns** [comment]
> Yes, running directly means executing macOS code;

I am using a mac, so I hopefully can figure out whats done in sign processing. Can you give me a reference where the signing process is done on mac?

> I  patched the relevant calls. In fact, the 64-bit assembly code is platform-independent

Where do I find this code?

---

Yesterday I may have found out some information. The response from `/lcdn/session` seems to be an `ASN.1 - X.690 - DER` certificate. Maybe it is used to sign the header from `register` (like JWT). The header could therefore consist of machine-specific information such as serial number, device, etc.

Maybe theese two links are helpful for further investigations:
- https://stackoverflow.com/questions/62638243/i-have-private-key-in-text-file-how-to-generate-pem-file-or-cer-file
- https://stackoverflow.com/questions/22743415/what-are-the-differences-between-pem-cer-and-der

I think we can manage that. Apple won't do any rocket science. here. We just need an approach ...
